### ISSUE TITLE
Failed to start disk device "X0"

### ISSUE THREAD

**Issue by randombenj** [lxc/lxd#9699]
<!--
Github issues are used for bug reports. For support questions, please use [our forum](https://discuss.linuxcontainers.org).

Please fill the template below as it will greatly help us track down your issue and reproduce it on our side.  
Feel free to remove anything which doesn't apply to you and add more information where it makes sense.
-->

# Required information

 * Distribution: Ubuntu
 * Distribution version: 20.04
 * The output of "lxc info" or if that fails:

```yaml
config:
  images.auto_update_interval: "1"
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- snapshot_schedule_aliases
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
- container_syscall_intercept
- container_disk_shift
- storage_shifted
- resources_infiniband
- daemon_storage
- instances
- image_types
- resources_disk_sata
- clustering_roles
- images_expiry
- resources_network_firmware
- backup_compression_algorithm
- ceph_data_pool_name
- container_syscall_intercept_mount
- compression_squashfs
- container_raw_mount
- container_nic_routed
- container_syscall_intercept_mount_fuse
- container_disk_ceph
- virtual-machines
- image_profiles
- clustering_architecture
- resources_disk_id
- storage_lvm_stripes
- vm_boot_priority
- unix_hotplug_devices
- api_filtering
- instance_nic_network
- clustering_sizing
- firewall_driver
- projects_limits
- container_syscall_intercept_hugetlbfs
- limits_hugepages
- container_nic_routed_gateway
- projects_restrictions
- custom_volume_snapshot_expiry
- volume_snapshot_scheduling
- trust_ca_certificates
- snapshot_disk_usage
- clustering_edit_roles
- container_nic_routed_host_address
- container_nic_ipvlan_gateway
- resources_usb_pci
- resources_cpu_threads_numa
- resources_cpu_core_die
- api_os
- container_nic_routed_host_table
- container_nic_ipvlan_host_table
- container_nic_ipvlan_mode
- resources_system
- images_push_relay
- network_dns_search
- container_nic_routed_limits
- instance_nic_bridged_vlan
- network_state_bond_bridge
- usedby_consistency
- custom_block_volumes
- clustering_failure_domains
- resources_gpu_mdev
- console_vga_type
- projects_limits_disk
- network_type_macvlan
- network_type_sriov
- container_syscall_intercept_bpf_devices
- network_type_ovn
- projects_networks
- projects_networks_restricted_uplinks
- custom_volume_backup
- backup_override_name
- storage_rsync_compression
- network_type_physical
- network_ovn_external_subnets
- network_ovn_nat
- network_ovn_external_routes_remove
- tpm_device_type
- storage_zfs_clone_copy_rebase
- gpu_mdev
- resources_pci_iommu
- resources_network_usb
- resources_disk_address
- network_physical_ovn_ingress_mode
- network_ovn_dhcp
- network_physical_routes_anycast
- projects_limits_instances
- network_state_vlan
- instance_nic_bridged_port_isolation
- instance_bulk_state_change
- network_gvrp
- instance_pool_move
- gpu_sriov
- pci_device_type
- storage_volume_state
- network_acl
- migration_stateful
- disk_state_quota
- storage_ceph_features
- projects_compression
- projects_images_remote_cache_expiry
- certificate_project
- network_ovn_acl
- projects_images_auto_update
- projects_restricted_cluster_target
- images_default_architecture
- network_ovn_acl_defaults
- gpu_mig
- project_usage
- network_bridge_acl
- warnings
- projects_restricted_backups_and_snapshots
- clustering_join_token
- clustering_description
- server_trusted_proxy
- clustering_update_cert
- storage_api_project
- server_instance_driver_operational
- server_supported_storage_drivers
- event_lifecycle_requestor_address
- resources_gpu_usb
- clustering_evacuation
- network_ovn_nat_address
- network_bgp
- network_forward
- custom_volume_refresh
- network_counters_errors_dropped
- metrics
- image_source_project
- clustering_config
- network_peer
- linux_sysctl
- network_dns
- ovn_nic_acceleration
- certificate_self_renewal
- instance_project_move
- storage_volume_project_move
- cloud_init
- network_dns_nat
- database_leader
- instance_all_projects
- clustering_groups
- ceph_rbd_du
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIICGDCCAZ+gAwIBAgIRAIG0NsaRld9U4RCr/jM92EEwCgYIKoZIzj0EAwMwOzEc
    MBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEbMBkGA1UEAwwScm9vdEBycm91
    d3BybGMwMDExMB4XDTIxMDMyMzA5NDc1NloXDTMxMDMyMTA5NDc1NlowOzEcMBoG
    A1UEChMTbGludXhjb250YWluZXJzLm9yZzEbMBkGA1UEAwwScm9vdEBycm91d3By
    bGMwMDExMHYwEAYHKoZIzj0CAQYFK4EEACIDYgAEe42kRWP6nNpKTAcQSelThK4x
    B9Kp8Cj1wnCBQ3jpzH27e+BoyFOvtnhM18SAzNjMVKA/Z6xy5vJpOd6rrBTFElPN
    XG3BS2DDrc5+t7vFUWJO+rf5arXBJEN/d0BJKDYXo2cwZTAOBgNVHQ8BAf8EBAMC
    BaAwEwYDVR0lBAwwCgYIKwYBBQUHAwEwDAYDVR0TAQH/BAIwADAwBgNVHREEKTAn
    gg1ycm91d3BybGMwMDExhwR/AAABhxAAAAAAAAAAAAAAAAAAAAABMAoGCCqGSM49
    BAMDA2cAMGQCMEmZCS6ki7UR46H2bKxLrxq6+xfNCZ9WjEJYTbaUyNa7W7xMUaBx
    6FirTubYAyffhQIwFRTsFLnyCh0Hc7A5qMMSMgQaoVCNFt0IlohMu3nTPzYfggXO
    HRXSGTnrBIo17eWf
    -----END CERTIFICATE-----
  certificate_fingerprint: b742be0f073a13465cada3a4ab3092b43a2db8900c4bb6198ede906a21601c61
  driver: lxc | qemu
  driver_version: 4.0.11 | 6.1.0
  firewall: xtables
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    seccomp_listener_continue: "true"
    shiftfs: "false"
    uevent_injection: "true"
    unpriv_fscaps: "true"
  kernel_version: 5.10.0-1052-oem
  lxc_features:
    cgroup2: "true"
    core_scheduling: "true"
    devpts_fd: "true"
    idmapped_mounts_v2: "true"
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    network_veth_router: "true"
    pidfd: "true"
    seccomp_allow_deny_syntax: "true"
    seccomp_notify: "true"
    seccomp_proxy_send_notify_fd: "true"
  os_name: Ubuntu
  os_version: "20.04"
  project: default
  server: lxd
  server_clustered: false
  server_name: rrouwprlc0011
  server_pid: 251159
  server_version: "4.21"
  storage: zfs
  storage_version: 2.0.2-1ubuntu5
  storage_supported_drivers:
  - name: ceph
    version: 15.2.14
    remote: true
  - name: btrfs
    version: 5.4.1
    remote: false
  - name: cephfs
    version: 15.2.14
    remote: true
  - name: dir
    version: "1"
    remote: false
  - name: lvm
    version: 2.03.07(2) (2019-11-30) / 1.02.167 (2019-11-30) / 4.43.0
    remote: false
  - name: zfs
    version: 2.0.2-1ubuntu5
    remote: false
```

# Issue description

To start GUI applications in lxd containers we mount the `/tmp/.X11-unix/X0` file to the container:

```sh
lxc launch ubuntu:20.04 c1
lxc config device add c1 X disk path=/tmp/.X11-unix/X0 source=/tmp/.X11-unix/X0
```

However since version `4.21` this fails with the following error:

```
Error: Failed to start device "X": Failed opening source path "/var/lib/snapd/hostfs/tmp/.X11-unix/X0": open /var/lib/snapd/hostfs/tmp/.X11-unix/X0: no such device or address
```

This used to work in version `4.20` and is reproducible on multiple machines and also on ubuntu 18.04.

Looking at `/var/lib/snapd/hostfs` there are no files:

```
$ ll /var/lib/snapd/hostfs                                                                                                                                                                                                                                
total 0
```

# Steps to reproduce

```sh
lxc launch ubuntu:20.04 c1
lxc config device add c1 X disk path=/tmp/.X11-unix/X0 source=/tmp/.X11-unix/X0
```

# Information to attach

 - [ ] ~~Any relevant kernel output (`dmesg`)~~
 - [ ] ~~Container log (`lxc info NAME --show-log`)~~
 - [ ] Container configuration (`lxc config show NAME --expanded`)
```yaml
architecture: x86_64
config:
  image.architecture: amd64
  image.description: ubuntu 20.04 LTS amd64 (release) (20211129)
  image.label: release
  image.os: ubuntu
  image.release: focal
  image.serial: "20211129"
  image.type: squashfs
  image.version: "20.04"
  volatile.base_image: a8402324842148ccfcbacbc69bf251baa9703916593089f0609e8d45e3185bff
  volatile.eth0.host_name: veth48e111e9
  volatile.eth0.hwaddr: 00:16:3e:1f:d3:9d
  volatile.idmap.base: "0"
  volatile.idmap.current: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.idmap.next: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.last_state.power: RUNNING
  volatile.uuid: a970d8cd-af00-494b-997b-098bb5a669b4
devices:
  eth0:
    name: eth0
    network: lxdbr0
    type: nic
  root:
    path: /
    pool: default
    type: disk
ephemeral: false
profiles:
- default
stateful: false
description: ""
```

 - [ ] ~~Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)~~
 - [ ] ~~Output of the client with --debug~~
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

```
location: none
metadata:
  context:
    ip: '@'
    method: GET
    protocol: unix
    url: /1.0
    username: uid=83265
  level: dbug
  message: Handling API request
timestamp: "2021-12-15T13:18:42.049008537+01:00"
type: logging


location: none
metadata:
  context:
    ip: '@'
    method: GET
    protocol: unix
    url: /1.0/instances/c1
    username: uid=83265
  level: dbug
  message: Handling API request
timestamp: "2021-12-15T13:18:42.05083918+01:00"
type: logging


location: none
metadata:
  context:
    ip: '@'
    method: GET
    protocol: unix
    url: /1.0/events
    username: uid=83265
  level: dbug
  message: Handling API request
timestamp: "2021-12-15T13:18:42.053949587+01:00"
type: logging


location: none
metadata:
  context: {}
  level: dbug
  message: 'New event listener: 833c9c95-9d40-415f-85d5-1a48842870f7'
timestamp: "2021-12-15T13:18:42.054109059+01:00"
type: logging


location: none
metadata:
  context:
    ip: '@'
    method: PUT
    protocol: unix
    url: /1.0/instances/c1
    username: uid=83265
  level: dbug
  message: Handling API request
timestamp: "2021-12-15T13:18:42.054470167+01:00"
type: logging


location: none
metadata:
  context: {}
  level: dbug
  message: 'New task Operation: 779c91ef-ba16-4c2b-ad76-b16658d92e34'
timestamp: "2021-12-15T13:18:42.078390893+01:00"
type: logging


location: none
metadata:
  class: task
  created_at: "2021-12-15T13:18:42.056932954+01:00"
  description: Updating instance
  err: ""
  id: 779c91ef-ba16-4c2b-ad76-b16658d92e34
  location: none
  may_cancel: false
  metadata: null
  resources:
    containers:
    - /1.0/containers/c1
    instances:
    - /1.0/instances/c1
  status: Pending
  status_code: 105
  updated_at: "2021-12-15T13:18:42.056932954+01:00"
timestamp: "2021-12-15T13:18:42.078733274+01:00"
type: operation


location: none
metadata:
  context: {}
  level: dbug
  message: 'Started task operation: 779c91ef-ba16-4c2b-ad76-b16658d92e34'
timestamp: "2021-12-15T13:18:42.078763911+01:00"
type: logging


location: none
metadata:
  class: task
  created_at: "2021-12-15T13:18:42.056932954+01:00"
  description: Updating instance
  err: ""
  id: 779c91ef-ba16-4c2b-ad76-b16658d92e34
  location: none
  may_cancel: false
  metadata: null
  resources:
    containers:
    - /1.0/containers/c1
    instances:
    - /1.0/instances/c1
  status: Running
  status_code: 103
  updated_at: "2021-12-15T13:18:42.056932954+01:00"
timestamp: "2021-12-15T13:18:42.079033967+01:00"
type: operation


location: none
metadata:
  context:
    ip: '@'
    method: GET
    protocol: unix
    url: /1.0/operations/779c91ef-ba16-4c2b-ad76-b16658d92e34
    username: uid=83265
  level: dbug
  message: Handling API request
timestamp: "2021-12-15T13:18:42.080953858+01:00"
type: logging


location: none
metadata:
  context:
    device: X
    instance: c1
    instanceType: container
    project: default
    type: disk
  level: dbug
  message: Starting device
timestamp: "2021-12-15T13:18:42.083946467+01:00"
type: logging


location: none
metadata:
  context: {}
  level: dbug
  message: 'Failure for task operation: 779c91ef-ba16-4c2b-ad76-b16658d92e34: Failed
    to start device "X": Failed opening source path "/var/lib/snapd/hostfs/tmp/.X11-unix/X0":
    open /var/lib/snapd/hostfs/tmp/.X11-unix/X0: no such device or address'
timestamp: "2021-12-15T13:18:42.084309262+01:00"
type: logging


location: none
metadata:
  context: {}
  level: dbug
  message: 'Scheduler: container c1 changed: re-balancing'
timestamp: "2021-12-15T13:18:42.084382046+01:00"
type: logging


location: none
metadata:
  class: task
  created_at: "2021-12-15T13:18:42.056932954+01:00"
  description: Updating instance
  err: 'Failed to start device "X": Failed opening source path "/var/lib/snapd/hostfs/tmp/.X11-unix/X0":
    open /var/lib/snapd/hostfs/tmp/.X11-unix/X0: no such device or address'
  id: 779c91ef-ba16-4c2b-ad76-b16658d92e34
  location: none
  may_cancel: false
  metadata: null
  resources:
    containers:
    - /1.0/containers/c1
    instances:
    - /1.0/instances/c1
  status: Failure
  status_code: 400
  updated_at: "2021-12-15T13:18:42.056932954+01:00"
timestamp: "2021-12-15T13:18:42.084523603+01:00"
type: operation


location: none
metadata:
  context: {}
  level: dbug
  message: 'Event listener finished: 833c9c95-9d40-415f-85d5-1a48842870f7'
timestamp: "2021-12-15T13:18:42.086136733+01:00"
type: logging


location: none
metadata:
  context:
    listener: 833c9c95-9d40-415f-85d5-1a48842870f7
  level: dbug
  message: Disconnected event listener
timestamp: "2021-12-15T13:18:42.086111062+01:00"
type: logging
```

**Comment by tomponline** [comment]
Can you show output of:

```
sudo nsenter --mount=/run/snapd/ns/lxd.mnt -- ls -la /var/lib/snapd/hostfs/tmp/.X11-unix/X0
```

And

```
ls -la /tmp/.X11-unix/X0
```

**Comment by randombenj** [comment]
```sh
$ sudo nsenter --mount=/run/snapd/ns/lxd.mnt -- ls -la /var/lib/snapd/hostfs/tmp/.X11-unix/X0
srwxrwxrwx 1 83265 users 0 Dec 14 09:07 /var/lib/snapd/hostfs/tmp/.X11-unix/X0
```

```sh
$ sudo ls -la /tmp/.X11-unix/X0
srwxrwxrwx 1 fassbinb users 0 Dec 14 09:07 /tmp/.X11-unix/X0
```

**Comment by tomponline** [comment]
Does passing through `/tmp/.X11-unix/` work?

**Comment by randombenj** [comment]
Yeah that seems to work:

```sh
lxc config device add c1 X disk path=/tmp/.X11-unix/ source=/tmp/.X11-unix/   
Device X added to c1
```

**Comment by tomponline** [comment]
And is the file you wanted to pass accessible inside the container that way?

The error you are getting is from:

https://github.com/lxc/lxd/blob/master/lxd/device/disk.go#L1325-L1328

It is quite a basic function call using the Go stdlib, so its strange that LXD running inside the snap mount namespace cannot open it.

**Comment by randombenj** [comment]
Sadly it's not:

```
root@c1:~# ls -alh /tmp/.X11-unix/
total 11K
drwxrwxrwt  2 nobody nogroup 40 Dec 15 12:34 .
drwxrwxrwt 10 root   root    10 Dec 15 12:30 ..
```

Well it was a few seconds ago:

```
root@c1:~# ls /tmp/.X11-unix/
X0  X1024  X1025
```

Thats strange ... ?

**Comment by randombenj** [comment]
Okay it seems to not be an lxd problem after all:

```sh
sudo ls -la /tmp/.X11-unix/X0
ls: cannot access '/tmp/.X11-unix/X0': No such file or directory
```

**Comment by tomponline** [comment]
Where was that run from?

**Comment by tomponline** [comment]
So it seems to be coming and going on the host or just in the container?

**Comment by randombenj** [comment]
It's from my host, so that's strange :thinking:

**Comment by randombenj** [comment]
So after a fresh reboot the `X0` file is there again in the host and contaienr and `xeyes` work again!
![image](https://user-images.githubusercontent.com/5184499/146190957-1a1ef5e9-1f51-4880-8c80-976c4b8293aa.png)

So mounting the folder seems a good workaround. Mounting the file directly still fails:

```sh
$ lxc launch ubuntu:20.04 c4
Creating c4
Starting c4
$ lxc config device add c4 X disk path=/tmp/.X11-unix/X0 source=/tmp/.X11-unix/X0
Error: Failed to start device "X": Failed opening source path "/var/lib/snapd/hostfs/tmp/.X11-unix/X0": open /var/lib/snapd/hostfs/tmp/.X11-unix/X0: no such device or address
$ lxc config device add c4 X disk path=/tmp/.X11-unix/ source=/tmp/.X11-unix/ 
Device X added to c4
```

@tomponline do you have any idea why this might behave differently than in `4.20`?

**Comment by randombenj** [comment]
Just reproduce the mounting of the `/tmp/.X11-unix/` folder on all colleagues' workstation, and the behaviour is the same, after `lxc config device add ...` the directory is empty.

**Comment by tomponline** [comment]
> It's from my host, so that's strange thinking

Yeah if the file is coming and going that is going to be a problem.

**Comment by tomponline** [comment]
Any ideas @stgraber ?

**Comment by stgraber** [comment]
@tomponline sounds like the issue here is with our changes to the bind mount logic and the source file being a socket.

You can't open() a socket the normal way. You may be able to open it enough with O_PATH though.

@tomponline should be reproducible by creating a socket with `nc -U -l` and attempting to pass that in.

**Comment by tomponline** [comment]
@stgraber ah OK so the socket is the difference, will test it.

**Comment by tomponline** [comment]
Yep that got it:

```
nc -U -l /home/user/foo.soc
lxc config device add ovndb foo disk source=/home/user/foo.sock path=/root/foo.sock
Error: Failed to start device "foo": Failed opening source path "/home/user/foo.sock": open /home/user/foo.sock: no such device or address
```

**Comment by tomponline** [comment]
thanks @stgraber O_PATH works well, and I was able to remove the O_READONLY as its not needed for directories when opening with O_PATH (and is ignored anyway from what I read).


Testing on VMs now to check no regression there.

**Comment by tomponline** [comment]
VM disk tests seem fine with this.

**Comment by newinnovations** [comment]
How to install this fix asap?  Reverting to 4.20 does not work as the database format has changed and edge does not feature it yet.

**Comment by tomponline** [comment]
Its been merged so should be in the edge snap channel soon.
